src/audio/audio.cpp

```diff
@@ -125,28 +158,78 @@ Set the master output volume.
 
 @param[in] volume   the master volume (between 0 and 1)
 */
-void Audio::setOutputVolume(ALfloat volume)
+void Audio::setOutputVolume(qreal volume)
 {
     QMutexLocker locker(&audioLock);
 
-    volume = std::max(0.f, std::min(volume, 1.f));
+    volume = std::max(0.0, std::min(volume, 1.0));
 
-    alListenerf(AL_GAIN, volume);
+    alListenerf(AL_GAIN, static_cast<ALfloat>(volume));
     checkAlError();
 }
 
-ALfloat Audio::inputVolume()
+/**
+@brief The minimum gain value for an input device.
+
+@return minimum gain value in dB
+*/
+qreal Audio::minInputGain() const
 {
     QMutexLocker locker(&audioLock);
+    return d->minInputGain;
+}
+
+/**
+@brief Set the minimum allowed gain value in dB.
 
-    return inGain;
+@note Default is -30dB; usually you don't need to alter this value;
+*/
+void Audio::setMinInputGain(qreal dB)
+{
+    QMutexLocker locker(&audioLock);
+    d->minInputGain = dB;
 }
 
-void Audio::setInputVolume(ALfloat volume)
+/**
+@brief The maximum gain value for an input device.
+
+@return maximum gain value in dB
+*/
+qreal Audio::maxInputGain() const
 {
     QMutexLocker locker(&audioLock);
+    return d->maxInputGain;
+}
+
+/**
+@brief Set the maximum allowed gain value in dB.
 
-    inGain = std::max(0.f, std::min(volume, 1.f));
+@note Default is 30dB; usually you don't need to alter this value.
+*/
+void Audio::setMaxInputGain(qreal dB)
```

Comment:
What is the advantage of creating a whole new class to encapsulate mutable parameters that that are essentially only changed by this class? I feel like that makes the code less readable by the constant access to object `d`. Isn't holding the values as member fields of this class good enough? Comments and variable isolation can help with establishing that the values +-30 dB are defaults. You can also declare the defaults as `constexpr` and completely make them immutable, allowing modification only via code changes.

If the separate class approach is necessary in your opinion, please move the `setX` `getX` functions that relate to gain to that class, or else the `Audio::Private` class is merely a glorified `struct` whose functionality merely adds an additional layer of indirection to otherwise things that could have been member fields of `Audio`.

Another potential idea to reduce the size of the `Audio` class would be to declare a `AudioGainDevice` (or a name of similar sort) that defines an abstract/virtual class that provides functions for gain control with suitable defaults. The `Audio` class can then appropriately derive from it as a subclass. Access control can be done via private/protected fields in the base class.